ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Registry.hs

```diff
@@ -0,0 +1,333 @@
+                , dpsOutstandingFifo = StrictSeq.empty
+                , dpsObjectsOwtPool = Map.empty
+                }
+              dgsPeerStates
+        }
+
+    -- TODO: this function needs to be tested!
+    -- Issue: https://github.com/IntersectMBO/ouroboros-network/issues/5151
+    unregisterPeerGlobalState ::
+      DecisionGlobalState peerAddr objectId object ->
+      DecisionGlobalState peerAddr objectId object
+    unregisterPeerGlobalState
+      st@DecisionGlobalState
+        { dgsPeerStates
+        } =
+        st
+          { dgsPeerStates = Map.delete peerAddr dgsPeerStates
+          }
+
+decisionLogicThread ::
+  forall m peerAddr objectId object.
+  ( MonadDelay m
+  , MonadSTM m
+  , MonadFork m
+  , MonadIO m
+  , Ord peerAddr
+  , Ord objectId
+  ) =>
+  Tracer m (TraceDecisionLogic peerAddr objectId object) ->
+  Tracer m ObjectDiffusionCounters ->
+  ObjectPoolWriter objectId object m ->
+  DecisionPolicy ->
+  PeerDecisionChannelsVar m peerAddr objectId object ->
+  DecisionGlobalStateVar m peerAddr objectId object ->
+  m Void
+decisionLogicThread decisionTracer countersTracer ObjectPoolWriter{opwHasObject} decisionPolicy decisionChannelsVar globalStateVar = do
+  labelThisThread "ObjectDiffusionInbound.decisionLogicThread"
+  forever $ do
+    -- We rate limit the decision making process, it could overwhelm the CPU
+    -- if there are too many inbound connections.
+    threadDelay const_DECISION_LOOP_DELAY
+
+    rng <- initStdGen
+
+    -- TODO: can we make this whole block atomic?
+    -- because makeDecisions should be atomic with respect to reading the global state and
+    -- reading the previous decisions
+    (newDecisions, counters) <- atomically $ do
+      decisionsChannels <- readTVar decisionChannelsVar
+      prevDecisions <- traverse readTVar decisionsChannels
+      globalState <- readTVar globalStateVar
+      hasObject <- opwHasObject
+      let newDecisions =
+            makeDecisions
+              DecisionContext
+                { dcRng = rng
+                , dcHasObject = hasObject
+                , dcDecisionPolicy = decisionPolicy
+                , dcGlobalState = globalState
+                , dcPrevDecisions = prevDecisions
+                }
+
+      peerToChannel <- readTVar decisionChannelsVar
+      -- Pair decision channel with the corresponding decision
+      let peerToChannelAndDecision =
+            Map.intersectionWith
+              (,)
+              peerToChannel
+              newDecisions
+      -- Send the newDecisions to the corresponding peers
+      traverse_
+        (\(chan, decision) -> writeTVar chan decision)
+        peerToChannelAndDecision
+
+      -- Return values for tracing purposes
+      let counters = makeObjectDiffusionCounters globalState
+      return (newDecisions, counters)
+
+    traceWith decisionTracer (TraceDecisionLogicDecisionsMade newDecisions)
+    traceWith countersTracer counters
+
+-- `5ms` delay
```

Comment:
Could this parameter be integrated into `DecisionPolicy`?

Also, 5ms is potentially up to 200 times a second. Isn't this _way_ too often? I would think that maybe somewhere in the realm of 10~20 times a second would be enough, but just guessing.